src/QsCompiler/Transformations/CallGraph.cs

```diff
@@ -68,64 +160,144 @@ internal static bool TryCombineTypeResolutionsForTarget(QsQualifiedName target,
                     success = success && !conflictingResolutionExists;
+
+                    // Add the resolution to the current dictionary.
                     combinedBuilder[typeParam] = paramRes;
                 }
             }
 
-            combined = combinedBuilder.ToImmutable();
+            combinedResolutions = combinedBuilder.ToImmutable();
             return success;
         }
+
+        /// <summary>
+        /// Walker that collects all of the type parameter references for a given ResolvedType
+        /// and returns them as a HashSet.
+        /// </summary>
+        internal class GetTypeParameters : TypeTransformation<GetTypeParameters.TransformationState>
+        {
+            /// <summary>
+            /// Walks the given ResolvedType and returns all of the type parameters referenced.
+            /// </summary>
+            public static HashSet<TypeParameterName> Apply(ResolvedType res)
+            {
+                var walker = new GetTypeParameters();
+                walker.OnType(res);
+                return walker.SharedState.TypeParams;
+            }
+
+            internal class TransformationState
+            {
+                public HashSet<TypeParameterName> TypeParams = new HashSet<TypeParameterName>();
+            }
+
+            private GetTypeParameters() : base(new TransformationState(), TransformationOptions.NoRebuild)
+            {
+            }
+
+            private static TypeParameterName AsTypeResolutionKey(QsTypeParameter tp) => Tuple.Create(tp.Origin, tp.TypeName);
+
+            public override ResolvedTypeKind OnTypeParameter(QsTypeParameter tp)
+            {
+                this.SharedState.TypeParams.Add(AsTypeResolutionKey(tp));
+                return base.OnTypeParameter(tp);
+            }
+        }
+
+        /// <summary>
+        /// Walker that checks a given type parameter resolution to see if it constricts
+        /// the type parameter to another type parameter of the same callable.
+        /// </summary>
+        internal class ConstrictionCheck : TypeTransformation<ConstrictionCheck.TransformationState>
+        {
+            private readonly TypeParameterName typeParamName;
+
+            /// <summary>
+            /// Walks the given ResolvedType, typeParamRes, and returns true if there is a reference
+            /// to a different type parameter of the same callable as the given type parameter, typeParam.
+            /// Otherwise returns false.
+            /// </summary>
+            public static bool Apply(TypeParameterName typeParam, ResolvedType typeParamRes)
+            {
+                var walker = new ConstrictionCheck(typeParam);
+                walker.OnType(typeParamRes);
+                return walker.SharedState.IsConstrictive;
+            }
+
+            internal class TransformationState
+            {
+                public bool IsConstrictive = false;
```

Comment:
This could also be a `TypeTransformation<bool>`.

Reply:
I could, but please see my response to the similar comment.